keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+
+When the Application Controller detects a newly created or updated object it shall do the following.
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. If the Application's `Selector` matches the object, and if the object does not contain an `OwnerReference` to the 
+Application, the Application Controller shall add an  `OwnerReference` from the Application to the object.
+
+As there is no guarantee that an Application's components will be created prior to the Application, the Application 
+Controller will do the following upon creation of an Application.
+
+1. [Select the Application's components](#selecting-components).
+1. For all selected objects, the Application Controller will add an `OwnerReference` from the Application to the object.
+
+When an Application object is deleted, the Garbage Collector will detect that the owner of the Application's components 
+no longer exists at the API Server. This will trigger deletion of the children objects. Note that the Application 
+Controller does not establish an ownership relationship between an Application and its dependencies. This is by design.
+Also note that, if the Application is composed of Kubernetes primitives that, in turn, aggregate other objects, it is 
+sufficient to only add an OwnerReference to the top level object. For instance, deleting a Application containing a 
+Deployment will delete the Deployment's ReplicaSets and Pods.
+
+#### Installation Status Reporting
+The Application Controller will additionally update the `Status` of Application objects. When the Application Controller 
+detects newly created objects, for all objects other than Applications, it shall do the following.
+
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. If the Application's `Selector` matches the object, the Application Controller shall update the Applications status 
+to indicate that component is installed.
+
+For all Application objects, the Application Controller shall do the following upon creation.
+
+1. [Select the Application's components](#selecting-components).
+1. For each selected object, the Application Controller will update the Application object's `Status` to indicate that 
+the component is installed. If a component is not found, the Applications status will be updated to indicate that the 
+component is not installed.
+1. [Retrieve the Application's dependencies](#retrieving-dependencies).
+1. For each dependency, the Application controller will update the the Application object's `Status` to indicate the 
+installation status of the dependency.
+1. The Application Controller will then scan all installed Applications in the same namespace as the Application 
+under consideration. If any of those Applications list the Application under consideration in their `Spec.Dependencies`, 
+their `Status` will be updated to indicate that the dependency has been installed.
```

Comment:
Is there a failure case? What happens if one object is never Installed? What happens to the Application if 0% of the objects are installed (e.g. b/c RBAC prevents)? Does the Application itself have a threshold for indicating that it is successful or failed? I only see `Ready` in the Go type defined at the top of this doc, and that's related to health check (which can pass even when not all components are installed).

Reply:
UIs can examine the installed indication from the applications status. Infant mortality detection and automated rollback would be interesting future work, but we don't even do that for core Workloads yet.